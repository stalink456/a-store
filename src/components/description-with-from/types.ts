import { OptionShape } from "@alfalab/core-components/select";

export type OptionsTypes = Record<
  "colors" | "sizes" | "stickerNumbers",
  OptionShape
> | null;

export type DescriptionType = {
  id: number;
  preview: string;
  title: string;
  price: number;
  description: string;
  colors?: string[];
  sizes?: string[];
  stickerNumbers?: number[];
  availability?: boolean;
};
