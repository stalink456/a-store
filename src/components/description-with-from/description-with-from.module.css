.product_description {
  width: 500px;
  padding: 0 15px;
}

.product_description__text {
  margin-top: 50px;
}

@media (max-width: 1135px) {
  .product_description {
    width: 400px;
  }
}

@media (max-width: 1065px) {
  .product_description {
    width: 640px;
  }
}

@media (max-width: 660px) {
  .product_description {
    width: 100%;
  }
}
