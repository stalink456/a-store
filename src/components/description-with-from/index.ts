export { DescriptionWithForm } from './description-with-form';
