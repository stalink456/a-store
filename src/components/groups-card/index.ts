export { GroupsCard } from './groups-card';
