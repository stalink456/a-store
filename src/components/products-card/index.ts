export { ProductsCard } from "./products-card";
