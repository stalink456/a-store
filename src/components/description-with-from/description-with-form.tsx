import React from "react";
import { Space } from "@alfalab/core-components/space";
import { Typography } from "@alfalab/core-components/typography";
import { DescriptionType } from "./types";
import { ProductForm } from "components/product-form";

import styles from "./description-with-from.module.css";

export const DescriptionWithForm: React.FC<DescriptionType> = React.memo(
  (props) => {
    const {
      id,
      preview,
      title,
      price,
      colors,
      sizes,
      stickerNumbers,
      description,
      availability,
    } = props;

    const renderForm = () => {
      return availability ? (
        <ProductForm
          id={id}
          title={title}
          preview={preview}
          price={price}
          colors={colors}
          sizes={sizes}
          stickerNumbers={stickerNumbers}
        />
      ) : (
        <Typography.Title
          tag="h2"
          weight="medium"
          view="small"
          color="negative"
        >
          Товара нет в наличии
        </Typography.Title>
      );
    };

    return (
      <div className={styles.product_description}>
        <Space direction="vertical" fullWidth size={15}>
          <Typography.Title tag="h1" weight="medium" view="small">
            {title}
          </Typography.Title>

          <Typography.Title tag="h2" weight="bold" view="small">
            {price} ₽
          </Typography.Title>

          {renderForm()}

          <Space
            direction="vertical"
            className={styles.product_description__text}
          >
            <Typography.Text tag="p" weight="medium" view="primary-small">
              {description}
            </Typography.Text>
          </Space>
        </Space>
      </div>
    );
  }
);
