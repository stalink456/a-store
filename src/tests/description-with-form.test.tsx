import { MemoryRouter } from 'react-router-dom';
import { waitFor } from '@testing-library/react';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { DescriptionWithForm } from 'components/description-with-from';
import * as product from './mocks/get-product-200.json';

const mockedProduct = product['product'];

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
  useSelector: jest.fn(),
}));

describe('Description component', () => {
  it('should render Description component', () => {
    const { baseElement } = render(<DescriptionWithForm {...mockedProduct} />, {
      wrapper: MemoryRouter,
    });

    expect(baseElement).toMatchSnapshot();
  });

  it('should render Description component with title', async () => {
    render(<DescriptionWithForm {...mockedProduct} />, {
      wrapper: MemoryRouter,
    });

    const result = await screen.findByText('Футболка с бархатными стикерами');
    expect(result).toBeInTheDocument();
  });

  it('should render Description component with select color', async () => {
    render(<DescriptionWithForm {...mockedProduct} />, {
      wrapper: MemoryRouter,
    });

    await waitFor(() =>
      expect(screen.queryByText('red')).not.toBeInTheDocument()
    );

    fireEvent.click(screen.getByText('Выберите цвет'));
    fireEvent.click(screen.getByText('красный'));

    await waitFor(async () =>
      expect(await screen.findByText('красный')).toBeInTheDocument()
    );
  });

  it('should render Description component with select size', async () => {
    render(<DescriptionWithForm {...mockedProduct} />, {
      wrapper: MemoryRouter,
    });

    await waitFor(() =>
      expect(screen.queryByText('XS')).not.toBeInTheDocument()
    );

    fireEvent.click(screen.getByText('Выберите размер'));
    fireEvent.click(screen.getByText('XS'));

    await waitFor(async () =>
      expect(await screen.findByText('XS')).toBeInTheDocument()
    );
  });

  it('should render Description component with select stickerNumber', async () => {
    render(<DescriptionWithForm {...mockedProduct} />, {
      wrapper: MemoryRouter,
    });

    await waitFor(() =>
      expect(screen.queryByText('1')).not.toBeInTheDocument()
    );

    fireEvent.click(screen.getByText('Выберите номер стикера'));
    fireEvent.click(screen.getByText('1'));

    await waitFor(async () =>
      expect(await screen.findByText('1')).toBeInTheDocument()
    );
  });

  it('should render Description component with enabled button', async () => {
    render(<DescriptionWithForm {...mockedProduct} />, {
      wrapper: MemoryRouter,
    });
    await waitFor(() => expect(screen.getByTestId('basket')).toBeDisabled());

    fireEvent.click(screen.getByText('Выберите цвет'));
    fireEvent.click(screen.getByText('красный'));

    fireEvent.click(screen.getByText('Выберите размер'));
    fireEvent.click(screen.getByText('XS'));

    fireEvent.click(screen.getByText('Выберите номер стикера'));
    fireEvent.click(screen.getByText('1'));

    await waitFor(() =>
      expect(screen.getByTestId('basket')).not.toBeDisabled()
    );
  });

  it("shouldn't render Description component with non-existent title", async () => {
    render(<DescriptionWithForm {...mockedProduct} />, {
      wrapper: MemoryRouter,
    });

    const result = screen.queryByText('non-existent title');
    await waitFor(() => expect(result).not.toBeInTheDocument());
  });

  it("shouldn't render Description component with non-existent option", async () => {
    render(<DescriptionWithForm {...mockedProduct} />, {
      wrapper: MemoryRouter,
    });

    fireEvent.click(screen.queryByText('Выберите цвет') as HTMLElement);

    await waitFor(() =>
      expect(screen.queryByText('orange')).not.toBeInTheDocument()
    );
  });
});
