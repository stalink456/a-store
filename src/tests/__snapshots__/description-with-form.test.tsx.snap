// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Description component should render Description component 1`] = `
<body>
  <div>
    <div
      class="product_description"
    >
      <div
        class="space__spaceContainer_nr3as space__vertical_nr3as space__start_nr3as space__spaceContainerFullWidth_nr3as"
      >
        <div
          class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
          style="margin-bottom: 15px;"
        >
          <h1
            class="typography__component_19zgr typography__styrene-small_pxbut typography__medium_pxbut"
          >
            Футболка с бархатными стикерами
          </h1>
        </div>
        <div
          class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
          style="margin-bottom: 15px;"
        >
          <h2
            class="typography__component_19zgr typography__styrene-small_pxbut typography__bold_pxbut"
          >
            1799
             ₽
          </h2>
        </div>
        <div
          class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
          style="margin-bottom: 15px;"
        >
          <form
            class="form"
          >
            <div
              class="space__spaceContainer_nr3as space__vertical_nr3as space__start_nr3as"
            >
              <div
                class="space__spaceItem_nr3as"
              >
                <p
                  class="typography__paragraph_c16y5 typography__primary-medium_c16y5"
                >
                  Цвет
                </p>
              </div>
            </div>
            <div
              aria-haspopup="listbox"
              aria-owns="downshift-1-menu"
              class="select__component_1bbba select__block_1bbba"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="select__component_qhc2e"
              >
                <div
                  class="form-control__component_e23cr form-control__component_1xedm form-control__s_e23cr form-control__block_e23cr form-control__hasRightAddons_e23cr"
                >
                  <div
                    aria-controls="downshift-1-menu"
                    aria-labelledby="downshift-1-label"
                    class="select__field_qhc2e form-control__inner_e23cr form-control__inner_1xedm"
                    id="downshift-0-input"
                    tabindex="0"
                  >
                    <div
                      class="form-control__inputWrapper_e23cr"
                    >
                      <div
                        class="form-control__input_e23cr"
                      >
                        <div
                          class="select__contentWrapper_qhc2e"
                        >
                          <span
                            class="select__placeholder_qhc2e"
                          >
                            Выберите цвет
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="form-control__addons_e23cr form-control__rightAddons_e23cr"
                    >
                      <svg
                        class="select__arrow_8kop9 select__arrow_qhc2e"
                        fill="currentColor"
                        focusable="false"
                        height="24"
                        role="img"
                        viewBox="0 0 24 24"
                        width="24"
                      >
                        <path
                          d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="space__spaceContainer_nr3as space__vertical_nr3as space__start_nr3as"
            >
              <div
                class="space__spaceItem_nr3as"
              >
                <p
                  class="typography__paragraph_c16y5 typography__primary-medium_c16y5"
                >
                  Размер
                </p>
              </div>
            </div>
            <div
              aria-haspopup="listbox"
              aria-owns="downshift-3-menu"
              class="select__component_1bbba select__block_1bbba"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="select__component_qhc2e"
              >
                <div
                  class="form-control__component_e23cr form-control__component_1xedm form-control__s_e23cr form-control__block_e23cr form-control__hasRightAddons_e23cr"
                >
                  <div
                    aria-controls="downshift-3-menu"
                    aria-labelledby="downshift-3-label"
                    class="select__field_qhc2e form-control__inner_e23cr form-control__inner_1xedm"
                    id="downshift-2-input"
                    tabindex="0"
                  >
                    <div
                      class="form-control__inputWrapper_e23cr"
                    >
                      <div
                        class="form-control__input_e23cr"
                      >
                        <div
                          class="select__contentWrapper_qhc2e"
                        >
                          <span
                            class="select__placeholder_qhc2e"
                          >
                            Выберите размер
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="form-control__addons_e23cr form-control__rightAddons_e23cr"
                    >
                      <svg
                        class="select__arrow_8kop9 select__arrow_qhc2e"
                        fill="currentColor"
                        focusable="false"
                        height="24"
                        role="img"
                        viewBox="0 0 24 24"
                        width="24"
                      >
                        <path
                          d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="space__spaceContainer_nr3as space__vertical_nr3as space__start_nr3as"
            >
              <div
                class="space__spaceItem_nr3as"
              >
                <p
                  class="typography__paragraph_c16y5 typography__primary-medium_c16y5"
                >
                  Номер стикера
                </p>
              </div>
            </div>
            <div
              aria-haspopup="listbox"
              aria-owns="downshift-5-menu"
              class="select__component_1bbba select__block_1bbba"
              role="combobox"
              tabindex="-1"
            >
              <div
                class="select__component_qhc2e"
              >
                <div
                  class="form-control__component_e23cr form-control__component_1xedm form-control__s_e23cr form-control__block_e23cr form-control__hasRightAddons_e23cr"
                >
                  <div
                    aria-controls="downshift-5-menu"
                    aria-labelledby="downshift-5-label"
                    class="select__field_qhc2e form-control__inner_e23cr form-control__inner_1xedm"
                    id="downshift-4-input"
                    tabindex="0"
                  >
                    <div
                      class="form-control__inputWrapper_e23cr"
                    >
                      <div
                        class="form-control__input_e23cr"
                      >
                        <div
                          class="select__contentWrapper_qhc2e"
                        >
                          <span
                            class="select__placeholder_qhc2e"
                          >
                            Выберите номер стикера
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="form-control__addons_e23cr form-control__rightAddons_e23cr"
                    >
                      <svg
                        class="select__arrow_8kop9 select__arrow_qhc2e"
                        fill="currentColor"
                        focusable="false"
                        height="24"
                        role="img"
                        viewBox="0 0 24 24"
                        width="24"
                      >
                        <path
                          d="M12 14L6.5 8.5 5 10l7 7 7-7-1.5-1.5L12 14z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <button
              class="button__component_j4291 button__m_j4291 button__component_3xn9f button__primary_3xn9f button__block_j4291"
              data-testid="basket"
              disabled=""
              type="submit"
            >
              <span
                class="button__text_j4291 button__stretchText_j4291"
              >
                В корзину
              </span>
            </button>
          </form>
        </div>
        <div
          class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
        >
          <div
            class="space__spaceContainer_nr3as space__vertical_nr3as space__start_nr3as product_description__text"
          >
            <div
              class="space__spaceItem_nr3as"
            >
              <p
                class="typography__paragraph_c16y5 typography__primary-small_c16y5 typography__medium_c16y5"
              >
                Выберите стикер, а мы перенесём его на ткань — как на фото. Одежду можно стирать в машинке при 30 °C, вывернув наизнанку. Гладить — с внутренней стороны. Посмотреть и потрогать все стикеры можно в A-Store на Технопарке. А ещё там можно добавить сразу несколько стикеров на одну вещь.
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    alfa-portal-container=""
  />
</body>
`;
